remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,532 @@
+          break;
+        case VisitResult::NoProgress:
+          next_wl.insert(inst);
+          break;
+        case VisitResult::Error:
+          return;
+      }
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
+  }
+}
+
+VisitResult ForwardAliasVisitor::visitInstruction(llvm::Instruction &I) {
+  return VisitResult::Progress;
+}
+
+VisitResult ForwardAliasVisitor::visitAllocaInst(llvm::AllocaInst &I) {
+  LOG(INFO) << "Entered alloca instruction";
+  exclude.insert(&I);
+  LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+  return VisitResult::Progress;
+}
+
+// Visit a load instruction and update the alias map.
+VisitResult ForwardAliasVisitor::visitLoadInst(llvm::LoadInst &I) {
+  LOG(INFO) << "Entered load instruction";
+  // get the initial ptr
+  auto val = I.getPointerOperand();
+  // special case: loading the state ptr
+  if (exclude.count(val) && I.getType() != state_ptr->getType()) {
+    exclude.insert(&I);
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    return VisitResult::Progress;
+  } else {
+    auto ptr = offset_map.find(val);
+    if (ptr == offset_map.end()) {
+      return VisitResult::NoProgress;
+    } else {
+      // loads mean we now have an alias to the pointer
+      alias_map.insert({&I, ptr->second});
+      LOG(INFO) << "aliasing: " << LLVMThingToString(&I);
+      return VisitResult::Progress;
+    }
+  }
+}
+
+// Visit a store instruction and update the alias map.
+VisitResult ForwardAliasVisitor::visitStoreInst(llvm::StoreInst &I) {
+  LOG(INFO) << "Entered store instruction";
+  // get the initial ptr
+  auto val = I.getPointerOperand();
+  if (exclude.count(val)) {
+    exclude.insert(&I);
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    return VisitResult::Progress;
+  } else {
+    auto ptr = offset_map.find(val);
+    if (ptr == offset_map.end()) {
+      return VisitResult::NoProgress;
+    } else {
+      // loads mean we now have an alias to the pointer
+      alias_map.insert({&I, ptr->second});
+      LOG(INFO) << "aliasing: " << LLVMThingToString(&I);
+      return VisitResult::Progress;
+    }
+  }
+}
+
+// Visit a GEP instruction and update the offset map.
+VisitResult ForwardAliasVisitor::visitGetElementPtrInst(llvm::GetElementPtrInst &I) {
+  LOG(INFO) << "Entered GEP instruction";
+  // get the initial ptr
+  auto val = I.getPointerOperand();
+  if (exclude.count(val)) {
+    exclude.insert(&I);
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    return VisitResult::Progress;
+  } else {
+    auto ptr = offset_map.find(val);
+    if (ptr == offset_map.end()) {
+      // no pointer found
+      return VisitResult::NoProgress;
+    } else {
+      // get the offset
+      llvm::APInt offset;
+      if (I.accumulateConstantOffset(*dl, offset)) {
+        // use offset (getRawData extracts the uint64_t)
+        offset_map[&I] = *(ptr->second + offset.getRawData());
```

Comment:
This doesn't make sense. `getRawData` returns a pointer to the data (memory backing an arbitrary sized integer), then you are displacing and dereferencing it, which is equivalent to indexing into an array. `ptr->second` could be pretty big, and who knows where in memory the APInt's internal storage points, so this is really a buffer overflow that reads arbitrary memory. Instead, what you want is this:

```c++
offset_map[&I] = ptr->second + offset.getZExtValue();
```